vm-migration/src/protocol.rs

```diff
@@ -0,0 +1,231 @@
+// Copyright © 2020 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use crate::MigratableError;
+
+// Migration protocol
+// 1: Source establishes communication with destination (file socket or TCP connection.)
+// (The establishment is out of scope.)
+// 2: Source -> Dest : send "start command"
+// 3: Dest -> Source : sends "ok response" when read to accept state data
+// 4: Source -> Dest : sends "state command" followed by state data, length
+//                     in command is length of state data
+// 5: Dest -> Source : sends "ok response" when ready to accept memory data
+// 6: Source -> Dest : send "memory command" followed by table of u64 pairs (GPA, size)
+//                     followed by the memory described in those pairs.
+//                     !! length is size of table i.e. 16 * number of ranges !!
+// 7: Dest -> Source : sends "ok response" when ready to accept more memory data
+// 8..(n-2): Repeat steps 6 and 7 until source has no more memory to send
+// (n-1): Source -> Dest : send "stop command"
```

Comment:
> This is a simple protocol, very nice. I'm trying to think about live migration and how this protocol would/could be extended to support it, and I think it could get us quite far already. A few questions:
> 
>     * After a stop command, is it left to the destination to decide if it should start the new VM? In other words, does receiving the `stop` mean the source is definitely done sending memory/state over the wire?
> 

I was thinking about renaming "Stop" to "Complete". This is to signify that the migration is done (from the sender's perspective.)

>     * Is it out of spec to send a `state` command after a `memory` command, if e.g. the source VM's state has changed while sending the memory ranges?
> 

Currently we don't handle it well. But with live migration a later state will go through the `.restore()` only path vs creating a new VM from the snapshot. To allow the final state of the VM to be sent through.

>     * On steps 8..(n-2), is the multi-step process designed for passing memory deltas? The data structure would allow for it at least, which is nice.

Exactly. This is the intention. Multiple memory commands are for the subsequent live migration state.